File: graph/dag/topological_sort.py
import sys
from collections import deque

# Fast I/O
input = sys.stdin.readline

class TopologicalSort:
    """
    Topological Sort (Kahn's Algorithm)
    Works on DAG (Directed Acyclic Graph)
    Time Complexity: O(V + E)
    """
    def __init__(self, n, adj):
        self.n = n
        self.adj = adj
        self.in_degree = [0] * (n + 1)
        self.compute_indegree()
        
    def compute_indegree(self):
        for u in range(1, self.n + 1):
            for v in self.adj[u]:
                self.in_degree[v] += 1
                
    def solve_kahn(self):
        """Returns topological ordering or [] if cycle detected"""
        q = deque()
        for i in range(1, self.n + 1):
            if self.in_degree[i] == 0:
                q.append(i)
        
        topo_order = []
        while q:
            # Use heapq here if you need lexicographically smallest sort
            u = q.popleft()
            topo_order.append(u)
            
            for v in self.adj[u]:
                self.in_degree[v] -= 1
                if self.in_degree[v] == 0:
                    q.append(v)
                    
        if len(topo_order) < self.n:
            return [] # Cycle detected
            
        return topo_order

def solve():
    try:
        line1 = input().split()
        if not line1: return
        n, m = map(int, line1)
        
        adj = [[] for _ in range(n + 1)]
        for _ in range(m):
            u, v = map(int, input().split())
            adj[u].append(v)
            # Directed graph
        
        topo = TopologicalSort(n, adj)
        result = topo.solve_kahn()
        
        if not result:
            print("IMPOSSIBLE") # Cycle
        else:
            print(*result)
            
    except ValueError:
        pass

if __name__ == "__main__":
    solve()

File: graph/dag/topological_sort.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * Topological Sort (Kahn's Algorithm)
 * Time Complexity: O(V + E)
 */

const int MAXN = 200005;
vector<int> adj[MAXN];
int in_degree[MAXN];

vector<int> kahn(int n) {
    queue<int> q;
    for (int i = 1; i <= n; i++) {
        if (in_degree[i] == 0) {
            q.push(i);
        }
    }
    
    vector<int> topo_order;
    while (!q.empty()) {
        // Use priority_queue<int, vector<int>, greater<int>> for lexicographically smallest
        int u = q.front();
        q.pop();
        topo_order.push_back(u);
        
        for (int v : adj[u]) {
            in_degree[v]--;
            if (in_degree[v] == 0) {
                q.push(v);
            }
        }
    }
    
    if (topo_order.size() < n) {
        return {}; // Cycle detected
    }
    
    return topo_order;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n, m;
    if (!(cin >> n >> m)) return 0;
    
    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        in_degree[v]++;
    }
    
    vector<int> result = kahn(n);
    
    if (result.empty()) {
        cout << "IMPOSSIBLE\n";
    } else {
        for (int i = 0; i < n; i++) {
            cout << result[i] << (i == n - 1 ? "" : " ");
        }
        cout << "\n";
    }
    
    return 0;
}
